### PR TITLE
8274083: Update testing docs to mention tiered testing

### PR DESCRIPTION
**Pull request by shipilev**
Now that OpenJDK has more or less complete `tier{1,2,3,4}` definitions, let's mention them in `testing.md`. 

Current patch is my braindump, I am open for suggestions :)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274083](https://bugs.openjdk.java.net/browse/JDK-8274083): Update testing docs to mention tiered testing

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to efd288046541de6309c10899880044b1f0e216be
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - **Reviewer**) ⚠️ Review applies to efd288046541de6309c10899880044b1f0e216be
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to cf2bda57b307a52ceb565166b6b09e7b13c5efce
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 32e9b5b4afcf0174c32d9984a3fdf582b9f742d5
 * [Mikhailo Seledtsov](https://openjdk.java.net/census#mseledtsov) (@mseledts - Committer) ⚠️ Review applies to efd288046541de6309c10899880044b1f0e216be
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5615/head:pull/5615` \
`$ git checkout pull/5615`

Update a local copy of the PR: \
`$ git checkout pull/5615` \
`$ git pull https://git.openjdk.java.net/jdk pull/5615/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5615`

View PR using the GUI difftool: \
`$ git pr show -t 5615`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5615.diff">https://git.openjdk.java.net/jdk/pull/5615.diff</a>

</details>

### CODE REVIEW

File: doc/testing.html [hunk 1]
```diff
@@ -67,6 +68,15 @@ <h2 id="test-selection">Test selection</h2>
...
+<ul>
+<li><p><code>tier1</code>: This test group is the first line of defense against bugs. Multiple developers run these tests every day. Normally, at least this tier is ran before integration. Because of the widespread, the tests in <code>tier1</code> are carefully selected and optimized to run fast, and to run in the most stable manner. The test failures in <code>tier1</code> are usually followed up on quickly, either with fixes, or adding relevant tests to problem list. <a href="../.github/workflows/">GitHub Actions workflows</a>, if enabled, run <code>tier1</code> tests.</p></li>
+<li><p><code>tier2</code>: This test group covers even more ground. These contain, among other things, tests that either run for too long to be at <code>tier1</code>, test unstable and/or experimental features, test less essential JDK components.</p></li>
+<li><p><code>tier3</code>: This test group covers more stressful tests, the tests for corner cases not covered by previous tiers, plus the tests that require GUIs. As such, this suite should either be run with low concurrency (<code>TEST_JOBS=1</code>), or without headful tests (<code>JTREG_KEYWORDS=\!headful</code>), or both.</p></li>
```
Comment (AlanBateman) [inline doc/testing.html:77]: It's not clear to me that these descriptions are useful. Instead of "First line of defense against bugs" then it might be better to say something about the hotspot and core library tests that are typically run in tier1. Folks contributing to some areas of the core, I/O, and networking libraries will need to learn about tier2 as that is where many of the tests for these APIs and implementations are. So I wouldn't use words like "unstable", "experimental", "less essential" but maybe say that it includes tests that may require some configuration, like changing firewall rules to allow multicast or other traffic.

Comment (dholmes-ora) [reply doc/testing.html:77]: Ideal world: run every test on every platform before every integration.
Real world: run as many tests as is practical having consideration for time of testing and the cost of testing resources, and the scope of the change.

It is assumed/expected you will have run all tests pertaining to the area of code change, that can reasonably be expected to be run.

The tiered testing is intended to give broader coverage to ensure no unintended consequences of a change. At a minimum run all tier1 tests on all non-trivial integrations. If you have time and resources, and think a change may have broader impact than what has been explicitly tested, then run further tiers.

Comment (shipilev) [reply doc/testing.html]: All right, thanks for the inputs. Let me try again, see new commit.

File: doc/testing.html [hunk 2]
```diff
@@ -67,6 +68,18 @@ <h2 id="test-selection">Test selection</h2>
...
+<h3 id="common-test-groups">Common Test Groups</h3>
+<p>In the ideal world, contributors would be advised to run all the tests for every change. But in real world, one could only be expected to run as many tests as practical, while being mindful of the scope for the change, the testing resources available, etc.</p>
+<p>The source tree currently defines a few common test groups in the relevant <code>TEST.groups</code> files. The test groups are either covering a specific component, for example <code>hotspot_gc</code>. It is a good idea to look into <code>TEST.groups</code> files to get a sense what tests are relevant to a particular JDK component.</p>
```
Comment (dholmes-ora) [inline doc/testing.html:73]: You state "The test groups are either ..." but there is no "or ..." in that sentence.

**Review by dholmes-ora — APPROVED** [review]
Hi Aleksey,

I don't have any significant misgivings about this version - thanks.

One correction needed below.

Thanks,
David

File: doc/testing.html (hunk 2 above)
Comment (shipilev) [reply doc/testing.html]: Thanks, see new commit.

File: doc/testing.md [hunk 3]
```diff
@@ -64,6 +64,51 @@ jtreg:$(TOPDIR)/test/nashorn:tier1 jtreg:$(TOPDIR)/test/jaxp:tier1`. You can
...
+- `tier2`: This test group covers even more ground. These contain, among other things,
+tests that either run for too long to be at `tier1`, tests for less stable and/or experimental
+features, tests for less essential JDK components (for example, jaxp).
+
+- `tier3`: This test group includes more stressful tests, the tests for corner cases
+not covered by previous tiers, plus the tests that require GUIs. As such, this suite
+should either be run with low concurrency (`TEST_JOBS=1`), or without headful tests
+(`JTREG_KEYWORDS=\!headful`), or both.
+
+- `tier4`: This test group includes every other test not covered by previous tiers. It includes,
```
Comment (DamonFool) [inline doc/testing.md:107]: `other test` => `other tests`

Comment (shipilev) [reply doc/testing.md:107]: No, I think "every other X" puts "X" in singular form: https://www.merriam-webster.com/dictionary/every%20other.

Comment (DamonFool) [reply doc/testing.md:107]: > No, I think "every other X" puts "X" in singular form: https://www.merriam-webster.com/dictionary/every%20other.

Okay, I agree.
Thanks.

**Review by DamonFool — APPROVED** [review]
LGTM.

Thanks for this kind of doc update.

**Review by tstuefe — APPROVED** [review]
I like this. Thanks for doing this (and for the actual definition of the tiers). This hopefully makes communicating about test coverage much easier.

As I understood, tier1-3 are the same as Oracle tier1-3?

File: doc/testing.md [hunk 4]
```diff
@@ -64,6 +64,51 @@ jtreg:$(TOPDIR)/test/nashorn:tier1 jtreg:$(TOPDIR)/test/jaxp:tier1`. You can
...
+special configuration. When choosing which tiers to run, contributors are advised to run the
+first N tiers they can afford to run.
+
+A brief description of the tiered test groups:
+
+- `tier1`: This is the lowest test tier. Multiple developers run these tests every day.
+Normally, at least this tier should be clean before integration. Because of the widespread use,
+the tests in `tier1` are carefully selected and optimized to run fast, and to run in the
+most stable manner. The test failures in `tier1` are usually followed up on quickly, either
+with fixes, or adding relevant tests to problem list. [GitHub Actions workflows](../.github/workflows/),
```
Comment (magicus) [inline doc/testing.md:95]: I don't think the link to the `.github` directory makes sense here. If there should be a link at all, it should point to information on how GHA works and how to enable these. And frankly I'm not sure if any such comprehensive documentation exists. :-( Perhaps in the Skara wiki? Otherwise the link is just distracting.

**Review by magicus — CHANGES_REQUESTED** [review]
I think this looks overall good. Just do something about that GHA link and I'm perfectly happy! :)

File: doc/testing.md (hunk 4 above)
Comment (shipilev) [reply doc/testing.md]: Sure, I actually put the link there very optimistically :) Dropped it in the new commit.

**Review by magicus — APPROVED** [review]

File: doc/testing.html [hunk 5]
```diff
@@ -67,6 +68,18 @@ <h2 id="test-selection">Test selection</h2>
...
 <p>The test specifications given in <code>TEST</code> is parsed into fully qualified test descriptors, which clearly and unambigously show which tests will be run. As an example, <code>:tier1</code> will expand to <code>jtreg:$(TOPDIR)/test/hotspot/jtreg:tier1 jtreg:$(TOPDIR)/test/jdk:tier1 jtreg:$(TOPDIR)/test/langtools:tier1 jtreg:$(TOPDIR)/test/nashorn:tier1 jtreg:$(TOPDIR)/test/jaxp:tier1</code>. You can always submit a list of fully qualified test descriptors in the <code>TEST</code> variable if you want to shortcut the parser.</p>
+<h3 id="common-test-groups">Common Test Groups</h3>
+<p>In the ideal world, contributors would be advised to run all the tests for every change. But in real world, one could only be expected to run as many tests as practical, while being mindful of the scope for the change, the testing resources available, etc.</p>
```
[the remark this answers is not in the document]
Comment (shipilev) [reply doc/testing.html]: Added a version of that, thanks.

File: doc/testing.html [hunk 6]
```diff
@@ -67,6 +68,18 @@ <h2 id="test-selection">Test selection</h2>
...
+<p>A brief description of the tiered test groups:</p>
+<ul>
+<li><p><code>tier1</code>: This is the lowest test tier. Multiple developers run these tests every day. Normally, at least this tier should be clean before integration. Because of the widespread use, the tests in <code>tier1</code> are carefully selected and optimized to run fast, and to run in the most stable manner. The test failures in <code>tier1</code> are usually followed up on quickly, either with fixes, or adding relevant tests to problem list. GitHub Actions workflows, if enabled, run <code>tier1</code> tests.</p></li>
+<li><p><code>tier2</code>: This test group covers even more ground. These contain, among other things, tests that either run for too long to be at <code>tier1</code>, tests for less stable and/or experimental features, tests for less essential JDK components (for example, jaxp).</p></li>
```
[the remark this answers is not in the document]
Comment (shipilev) [reply doc/testing.html]: See new commit.

> Tests in tier2 need be run by anyone making changes in the security area, I/O, networking and many other core APIs. In general I think anyone changing code in the libraries area needs to look at TEST.groups to get some idea which tier or finer test group to run.

True. That is why the text says that component tests come first, and then you run first N tiers to capture anything else. If any component test is not in the component test group, and only in `tier2`, I'd say that is a test grouping bug, and the test should be in component test group to begin with. In other words, `tier2` testing does not replace the component group testing, and thus the requirement to run `tier2` if particular areas are changed would feel awkward to me.

>  Do we have tests for hotspot experimental features in hotspot_tier2?

We do move deprecating things out of `tier1` (see e.g. recent JDK-8272838).

**Review by mseledts — APPROVED** [review]
Change looks good to me, thank you.

**Review by dholmes-ora — APPROVED** [review]

**Review by DamonFool — APPROVED** [review]

File: doc/testing.html [hunk 7]
```diff
@@ -67,6 +68,19 @@ <h2 id="test-selection">Test selection</h2>
...
+<p>Contributors are expected to run the tests for the areas that are changed, and the first N tiers they can afford to run, but at least tier1.</p>
+<p>A brief description of the tiered test groups:</p>
+<ul>
+<li><p><code>tier1</code>: This is the lowest test tier. Multiple developers run these tests every day. Because of the widespread use, the tests in <code>tier1</code> are carefully selected and optimized to run fast, and to run in the most stable manner. The test failures in <code>tier1</code> are usually followed up on quickly, either with fixes, or adding relevant tests to problem list. GitHub Actions workflows, if enabled, run <code>tier1</code> tests.</p></li>
+<li><p><code>tier2</code>: This test group covers even more ground. These contain, among other things, tests that either run for too long to be at <code>tier1</code>, or may require special configuration, or tests that are less stable, or cover the broader range of less critical JVM and JDK features/components (for example, jaxp).</p></li>
```
Comment (AlanBateman) [inline doc/testing.html:80]: Thanks for the updates, I think it reads much better now.
A small suggestion is to change "less critical" to "non-core", and "jaxp" to "XML" as I don't expect too many people know what JAXP is.

**Review by AlanBateman — APPROVED** [review]

File: doc/testing.html (hunk 7 above)
Comment (shipilev) [reply doc/testing.html]: All right, changed. I'll be integrating this version soon.

**Review by AlanBateman — APPROVED** [review]

**Review by LanceAndersen — COMMENTED** [review]
Looks good to me.
